validate/main.go

```diff
@@ -0,0 +1,84 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"strings"
+
+	"github.com/coreos/ignition/config"
+	"github.com/coreos/ignition/internal/version"
+
+	"github.com/spf13/cobra"
+)
+
+var (
+	flagVersion bool
+	rootCmd     = &cobra.Command{
+		Use:   "ignvalidate [CONFIG.IGN]",
+		Short: "ignvalidate will validate Ignition configs",
+		Run:   runIgnValidate,
+	}
+)
+
+func main() {
+	rootCmd.Flags().BoolVar(&flagVersion, "version", false, "print the version of ignvalidate")
+	rootCmd.Execute()
+}
+
+func stdout(format string, a ...interface{}) {
+	fmt.Fprintf(os.Stdout, strings.TrimSpace(format)+"\n", a...)
+}
+
+func stderr(format string, a ...interface{}) {
+	fmt.Fprintf(os.Stderr, strings.TrimSpace(format)+"\n", a...)
+}
+
+func die(format string, a ...interface{}) {
+	stderr(format, a...)
+	os.Exit(1)
+}
+
+func runIgnValidate(cmd *cobra.Command, args []string) {
+	if flagVersion {
+		stdout(version.String)
+		return
+	}
+	if len(args) != 1 {
+		cmd.Usage()
+		os.Exit(1)
+	}
+	var in io.Reader
+	if args[0] == "-" {
+		in = os.Stdin
+	} else {
+		file, err := os.Open(args[0])
+		if err != nil {
+			die("couldn't open config: %v", err)
+		}
+		defer file.Close()
+		in = file
+	}
+	blob, err := ioutil.ReadAll(in)
+	if err != nil {
+		die("couldn't read config: %v", err)
+	}
+	cfg, rpt, err := config.Parse(blob)
+	if len(rpt.Entries) > 0 {
+		stdout(rpt.String())
+	}
+	if rpt.IsFatal() {
+		os.Exit(1)
+	}
+	if err != nil {
+		die("couldn't parse config: %v", err)
+	}
+	rpt = cfg.Validate()
```

Comment:
`Parse()` should already call `Validate()`, yes?

Reply:
It looks to me like it doesn't in the case of a version 1 config.

Reply:
Huh. I'd argue `ParseFromV1()` should call `validate.ValidateWithoutSource()` instead of calling Validate() here. Also, you want to call `validate.ValidateWithoutSource` not `configinstance.Validate()`. The latter is just the top level validator function that does global config checks (e.g. filesystem shadowing, files with missing filesystems).